File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
            res = []
            def dfs(index , temp , total):
                #base case
                if total == target:
                    res.append(temp.copy())
                    return
                if index >= len(nums) or total > target:
                    return

                temp.append(nums[index])
                dfs(index ,temp ,  total + nums[index])
                temp.pop()
                dfs(index+1 ,temp ,  total)

            dfs(0 ,[], 0)

            return res

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    
    void dfs(int index , vector<int>& temp , int total , int target , vector<int>& candidates, vector<vector<int>>& ans){

	if (total == target){
		ans.push_back(temp);
		return;
	}
	if (index >= (int)candidates.size() || total > target){
		return ;
	}

	temp.push_back(candidates[index]);
	dfs(index  , temp , total + candidates[index] , target , candidates , ans);

	temp.pop_back();
	dfs(index+1, temp , total , target , candidates , ans);
}
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        
	vector<vector<int>> ans ;
	vector<int> temp ;

	dfs(0 , temp , 0 , target , candidates , ans);

	return ans;

    }
};